libsolidity/analysis/DeclarationTypeChecker.cpp

```diff
@@ -140,6 +140,36 @@ bool DeclarationTypeChecker::visit(StructDefinition const& _struct)
 	return false;
 }
 
+bool DeclarationTypeChecker::visit(UserDefinedValueTypeDefinition const& _userDefined)
```

Comment:
Ok, I see why you did it this way: In `DeclarationTypeChecker::endVisit(UserDefinedTypeName`, the underlying type needs to be available, but I think this is a more serious problem and cannot be solved by changing the visitation order: We can also have forward references to other user-defined types at that point. I think the solution is the same as for struct types: Change the type constructor so that it only takes the AST definition, but do not analyze the type further.

Reply:
Indeed, the following code still triggers the assertion:
```
function f(MyIntB x) pure {}
type MyIntB is MyIntB;
```